### PR TITLE
boa: add generator support

### PR DESCRIPTION
**Pull request by WenheLI**
This PR adds generator support to align JS generator syntax

### CODE REVIEW

File: packages/boa/lib/proxy.js [hunk 1]
```diff
@@ -347,6 +347,11 @@ function _internalWrap(T, src={}) {
               `or one-level properties.`);
 
         if (typeof value === 'function') {
+          // the next called on generator expects its caller to be a `generator` type where
+          // proxy does not have such base class
+          if (name.toString() === 'next' && getTypeInfo(T).name === 'generator') {
```
Comment (yorkie) [inline packages/boa/lib/proxy.js:352]:
```suggestion
          if (type.module === 'builtins' && type.name === 'generator' && name === 'next') {
```

File: packages/boa/lib/proxy.js [hunk 2]
```diff
@@ -347,6 +347,11 @@ function _internalWrap(T, src={}) {
               `or one-level properties.`);
 
         if (typeof value === 'function') {
+          // the next called on generator expects its caller to be a `generator` type where
+          // proxy does not have such base class
```
Comment (yorkie) [inline packages/boa/lib/proxy.js:351]:
```suggestion
          // proxy does not have such base class.
```

File: packages/boa/lib/proxy.js [hunk 3]
```diff
@@ -347,6 +347,11 @@ function _internalWrap(T, src={}) {
               `or one-level properties.`);
 
         if (typeof value === 'function') {
+          // the next called on generator expects its caller to be a `generator` type where
```
Comment (yorkie) [inline packages/boa/lib/proxy.js:350]:
```suggestion
          // the `next` called on generator expects its caller to be a `generator` type where
```

File: packages/boa/lib/proxy.js (hunk 1 above)
Comment (yorkie) [reply packages/boa/lib/proxy.js:352]: Is there other functions that we need to change as well.

File: packages/boa/lib/proxy.js [hunk 4]
```diff
@@ -331,8 +331,8 @@ function _internalWrap(T, src={}) {
   });
 
   // Create the proxy object for handlers
-  let newTarget;
-  return (newTarget = new Proxy(src, {
```
Comment (yorkie) [inline packages/boa/lib/proxy.js:335]: Unnecessary changes.

File: packages/boa/src/core/object.cc [hunk 5]
```diff
@@ -42,6 +42,7 @@ Object PythonObject::Init(Napi::Env env, Object exports) {
           InstanceMethod("__getitem__", &PythonObject::GetItem),
           InstanceMethod("__setitem__", &PythonObject::SetItem),
           InstanceMethod("__delitem__", &PythonObject::DelItem),
+          InstanceMethod("__next__", &PythonObject::Next),
```
[the remark this answers is not in the document]
Comment (WenheLI) [reply packages/boa/src/core/object.cc:45]: This is a magic method for python `generator` we should implement it incase someone will call it.

File: packages/boa/lib/proxy.js (hunk 4 above)
Comment (WenheLI) [reply packages/boa/lib/proxy.js:335]: This is for clarity.

File: packages/boa/src/core/object.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (yorkie) [reply packages/boa/src/core/object.cc:45]: Actually, these are not corresponding to magic methods exactly, and someone could call any magic method on a Python object, these internal methods are only used internally.

File: packages/boa/lib/proxy.js (hunk 4 above)
Comment (yorkie) [reply packages/boa/lib/proxy.js:335]: But I don't see it's clarity more, and it's unrelated to this PR.

File: packages/boa/lib/proxy.js [hunk 6]
```diff
@@ -102,7 +102,7 @@ function asHandleObject(T) {
   };
 }
 
-function _internalWrap(T, src={}) {
+function _internalWrap(T, src={}, thisType={}) {
```
Comment (yorkie) [inline packages/boa/lib/proxy.js:105]:
```suggestion
function _internalWrap(T, src={}, pytype={}) {
```

File: packages/boa/lib/proxy.js [hunk 7]
```diff
@@ -347,6 +348,11 @@ function _internalWrap(T, src={}) {
               `or one-level properties.`);
 
         if (typeof value === 'function') {
+          // the `next` called on generator expects its caller to be a `generator` type where
+          // proxy does not have such base class.
+          if (thisType.module === 'builtins' && thisType.name === 'generator' && name === 'next') {
```
Comment (yorkie) [inline packages/boa/lib/proxy.js:353]:
```suggestion
          if (pytype.module === 'builtins' && pytype.name === 'generator' && name === 'next') {
```

File: packages/boa/src/core/object.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (yorkie) [reply packages/boa/src/core/object.cc:45]: Ditto, this could be removed.

File: docs/manual/intro-to-boa.md [hunk 8]
```diff
@@ -131,6 +131,32 @@ bytes.fromhex(foobar.toString('hex'));
 // "b'foobar'"
 ```
 
+### Use of Generator
```
Comment (yorkie) [inline docs/manual/intro-to-boa.md:134]: This should be prepended before https://github.com/alibaba/pipcook/blob/49b821201e61fb17dc63cef31806bec675b531c5/docs/manual/intro-to-boa.md#python-functions-in-worker_threads.

File: docs/manual/intro-to-boa.md [hunk 9]
```diff
@@ -131,6 +131,32 @@ bytes.fromhex(foobar.toString('hex'));
...
+
+`Boa` will also handle `Generator` class in the native python runtime. JavaScript syntax of `Generator` could be applied to Python `Generator` directly:
```
Comment (yorkie) [inline docs/manual/intro-to-boa.md:136]:
```suggestion
The package is able to handle the Python generator class in JavaScript directly:
```

File: docs/manual/intro-to-boa.md [hunk 10]
```diff
@@ -131,6 +131,32 @@ bytes.fromhex(foobar.toString('hex'));
...
+
+```js
+
```
Comment (yorkie) [inline docs/manual/intro-to-boa.md:139]: Remove this line.

File: docs/manual/intro-to-boa.md [hunk 11]
```diff
@@ -131,6 +131,32 @@ bytes.fromhex(foobar.toString('hex'));
...
+const generator = generatorFromPython(); // Generator
```
Comment (yorkie) [inline docs/manual/intro-to-boa.md:140]: I think it's better to represent it from Python:

- prepend a python block to write a generator function.
- import the generator object using boa.import here.

File: packages/boa/lib/delegators/builtins.generator.js [hunk 12]
```diff
@@ -0,0 +1,14 @@
+module.exports = (T, wrap) => {
+
```
Comment (yorkie) [inline packages/boa/lib/delegators/builtins.generator.js:2]: Unnecessary blank.

File: packages/boa/lib/delegators/builtins.generator.js [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+  function *generator() {
```
Comment (yorkie) [inline packages/boa/lib/delegators/builtins.generator.js:3]:
```suggestion
  function *generatorProxy() {
```

**Review by yorkie — CHANGES_REQUESTED** [review]

File: docs/manual/intro-to-boa.md [hunk 14]
```diff
@@ -274,6 +274,42 @@ In Node.js version < `v14.x`, you also need to add the [`--experimental-modules`
 $ node --experimental-modules --experimental-loader @pipcook/boa/esm/loader.mjs app.mjs
 ```
 
+### Python Generators in JS
```
Comment (yorkie) [inline docs/manual/intro-to-boa.md:277]:
```suggestion
### Generators
```

File: docs/manual/intro-to-boa.md [hunk 15]
```diff
@@ -274,6 +274,42 @@ In Node.js version < `v14.x`, you also need to add the [`--experimental-modules`
...
+
+The package is able to handle the Python generator class in JavaScript directly:
```
Comment (yorkie) [inline docs/manual/intro-to-boa.md:279]:
```suggestion
The package is able to handle the Python generators in JavaScript directly:
```

File: docs/manual/intro-to-boa.md [hunk 16]
```diff
@@ -274,6 +274,42 @@ In Node.js version < `v14.x`, you also need to add the [`--experimental-modules`
...
+# Write a Python Generator in count_down.py
+def count_down(count):
+  while count >= 0:
+    yield count
+    count -= 1
+```
+
+```js
+const boa = require('@pipcook/boa')
+const countDown = boa.import('path_to_above_py')
```
Comment (yorkie) [inline docs/manual/intro-to-boa.md:291]:
```suggestion
const countDown = boa.import('count_down')
```

File: docs/manual/intro-to-boa.md [hunk 17]
```diff
@@ -274,6 +274,42 @@ In Node.js version < `v14.x`, you also need to add the [`--experimental-modules`
...
+const generator = countDown(3); // Generator
+
+// You can access data via the following syntax
+for (const element of generator) {
+  console.log(element) // 3 2 1 0
+}
+
+// or use typical next syntax
+
+const generator = countDown(3); // Generator
```
Comment (yorkie) [inline docs/manual/intro-to-boa.md:301]: I think this example should be the first case, and it's good to split into different blocks.

File: docs/manual/intro-to-boa.md [hunk 18]
```diff
@@ -274,6 +274,42 @@ In Node.js version < `v14.x`, you also need to add the [`--experimental-modules`
...
+The package is able to handle the Python generator class in JavaScript directly:
+
+```python
+# Write a Python Generator in count_down.py
+def count_down(count):
+  while count >= 0:
+    yield count
+    count -= 1
+```
+
```
Comment (yorkie) [inline docs/manual/intro-to-boa.md:288]: We need to describe the following code block.

**Review by yorkie — APPROVED** [review]

**Review by FeelyChau — APPROVED** [review]

**Review by rickycao-qy — APPROVED** [review]
LGTM
